controllers/workflow_controller_test.go

```diff
@@ -0,0 +1,195 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/uuid"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	dwsv1alpha1 "github.com/HewlettPackard/dws/api/v1alpha1"
+)
+
+var _ = Describe("Workflow Controller Test", func() {
+
+	var (
+		wf *dwsv1alpha1.Workflow
+	)
+
+	BeforeEach(func() {
+		wfid := uuid.NewString()[0:8]
+		wf = &dwsv1alpha1.Workflow{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      wfid,
+				Namespace: corev1.NamespaceDefault,
+			},
+			Spec: dwsv1alpha1.WorkflowSpec{
+				DesiredState: dwsv1alpha1.StateProposal,
+				WLMID:        "test",
+				JobID:        0,
+				UserID:       0,
+				GroupID:      0,
+				DWDirectives: []string{},
+			},
+			Status: dwsv1alpha1.WorkflowStatus{
+				State:  dwsv1alpha1.StateProposal,
+				Ready:  false,
+				Status: dwsv1alpha1.StatusDriverWait,
+			},
+		}
+	})
+
+	AfterEach(func() {
+		if wf != nil {
+			Expect(k8sClient.Delete(context.TODO(), wf)).To(Succeed())
+
+			wfExpected := &dwsv1alpha1.Workflow{}
+			Eventually(func() error { // Delete can still return the cached object. Wait until the object is no longer present.
+				return k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(wf), wfExpected)
+			}).ShouldNot(Succeed())
+		}
+	})
+
+	It("Can complete Workflow driver states", func() {
+		action := "complete"
+		wf.Spec.DWDirectives = []string{
+			fmt.Sprintf("#DW STATUS action=%s", action),
+		}
+
+		driverStatus := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+		}
+
+		// Set a bunch of other statuses that should remain unchanged
+		statusWrongDriver := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   "notMyDriver",
+			DWDIndex:   0,
```

Comment:
This doesn't look realistic.  The DWDIndex refers the specific "#DW" directive found in the job script.   You have other states in the same driver status array that say the first (zeroth) one is $DRIVERID.  So the "notMyDriver" one would be a different #DW in the job script, and would have a different DWDIndex.

I wonder if the DWS webhook would have caught this.